### PR TITLE
feat!: preserve directory structure when globbing in config

### PR DESCRIPTION
**Pull request by digitalsadhu**
BREAKING CHANGE: directory structures are no longer flattened

Needed for Eik RFC-2 https://github.com/eik-lib/issues/issues/2

This PR introduces a new method to config called `.mappings()` which essentially reimplements (and is intended as a replacement for) the `pathsAndFiles` and `pathsAndFilesAbsolute` methods but conforms to the requirements laid out in RFC-2.

`.mappings()` returns an array of a new class instance of `FileMapping` which has `source` and `destination` properties. Source is an instance of a new class called `LocalFileLocation` and destination is an instance of a new class called `RemoteFileLocation`. These new classes make for very nice output from the `.mappings()` method, you get an array with a new output structure as follows:

```js
[
  FileMapping {
    source: LocalFileLocation {
      relative: 'folder/client.js',
      cwd: '/Users/richard.walker/Documents/src/@eik/common/test/fixtures',
      absolute: '/Users/richard.walker/Documents/src/@eik/common/test/fixtures/folder/client.js'
    },
    destination: RemoteFileLocation {
      packagePathname: '/pkg/pizza/0.0.0',
      filePathname: '/path/to/folder/client.js',
      url: [URL]
    }
  },
  FileMapping {
    source: LocalFileLocation {
      relative: 'folder/styles.css',
      cwd: '/Users/richard.walker/Documents/src/@eik/common/test/fixtures',
      absolute: '/Users/richard.walker/Documents/src/@eik/common/test/fixtures/folder/styles.css'
    },
    destination: RemoteFileLocation {
      packagePathname: '/pkg/pizza/0.0.0',
      filePathname: '/path/to/folder/styles.css',
      url: [URL]
    }
  }
]
```
([URL] above is a whatwg URL instance)

This PR also adds js doc type comments throughout the codebase which improves type hinting (at least in viscode)
A command `npm run schema:types` has been added to generate a ts definition file from the eik json json schema file. This can be run and the result committed if the json schema changes.

### CODE REVIEW

**Comment by digitalsadhu** [comment]
As of the latest commits, the new `config.mappings()` method supports all the defined examples in this comment here: https://github.com/eik-lib/issues/issues/2#issuecomment-779099732

File: lib/classes/eik-config.js [hunk 1]
```diff
@@ -1,19 +1,54 @@
...
+    val.endsWith('/') ? val.substr(0, val.length - 1) : val;
+
+/**
+ * @param {string} val
+ *
+ * @returns {string}
+ */
+const addTrailingSlash = (val) =>
+    val.endsWith('/') ? val : `${val}/`;
+
```
Comment (trygve-lie) [inline lib/classes/eik-config.js:34]: Are these methods intended to be used for when working on paths? Would using the node `join`  module where building paths solve this for instead of rolling our own? I mean, `join`  does normalize the slashes when joining paths.

Comment (digitalsadhu) [reply lib/classes/eik-config.js:34]: it doesn't normalise leading or trailing slashes at all unfortunately

Comment (digitalsadhu) [reply lib/classes/eik-config.js:34]:
```
> join('asd', 'asd', 'asd')
'asd/asd/asd'
> join('/asd', 'asd', 'asd')
'/asd/asd/asd'
> join('/asd', 'asd', 'asd/')
'/asd/asd/asd/'
> 
```

**Comment by digitalsadhu** [comment]
Updated PR description to better explain the code changes

File: lib/helpers/path-diff.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// @ts-check
+/* eslint-disable no-continue */
+const { basename, normalize } = require('path');
+const isGlob = require('is-glob');
+
+const leadingSlash = (val) => val.startsWith('/') ? val : `/${val}`;
```
Comment (trygve-lie) [inline lib/helpers/path-diff.js:6]: If we need this type of checks, would it make sense moving this and the two found [here](https://github.com/eik-lib/common/pull/103/files#diff-2a8a60d95f6c8a7f0bc3d7c30c319a2889895c59f7ed77dee91e33cf5a02eb03R24-R33) into a util module?

Comment (digitalsadhu) [reply lib/helpers/path-diff.js:6]: yeah, I likely would. Can take a refactor round and do some cleanup.

**Comment by trygve-lie** [comment]
> This PR also adds js doc type comments throughout the codebase which improves type hinting (at least in viscode)
> A command npm run schema:types has been added to generate a ts definition file from the eik json json schema file. This can be run and the result committed if the json schema changes.

Awesome. I am very happy to see this.

File: lib/helpers/path-diff.js (hunk 2 above)
Comment (trygve-lie) [reply lib/helpers/path-diff.js]: :+1:

**Comment by digitalsadhu** [comment]
Added type checking to the CI build

File: lib/helpers/path-diff.js (hunk 2 above)
Comment (digitalsadhu) [reply lib/helpers/path-diff.js]: Cleanup done

File: lib/classes/eik-config.js [hunk 3]
```diff
@@ -61,6 +120,36 @@ module.exports = class EikConfig {
...
     }
 
+    /**
+     * @returns {Promise<FileMapping[]>}
+     */
+    async mappings() {
+        const files = normalizeFilesDefinition(this.files);
+        const resolvedFiles = await resolveFiles(files, this.cwd);
+
+        return resolvedFiles.flatMap(([pathname, file, filePaths]) => {
```
Comment (gingermusketeer) [inline lib/classes/eik-config.js:130]: Would `destination`, `source` and `sourceFilePaths` be better names here?

File: lib/classes/eik-config.js [hunk 4]
```diff
@@ -82,7 +171,7 @@ module.exports = class EikConfig {
                 return [[pathname, filePaths[0], pathname, file]];
             }
             return filePaths.map((filePath) => [
-                path.join(pathname, path.basename(filePath)),
+                path.join(pathname, filePath),
```
Comment (gingermusketeer) [inline lib/classes/eik-config.js:174]: Do you want to change this function if it is being deprecated in favour of `mappings`? Should it be marked as deprecated?

File: lib/helpers/index.js [hunk 5]
```diff
@@ -2,5 +2,26 @@ const localAssets = require('./local-assets');
...
 const configStore = require('./config-store');
+const typeSlug = require('./type-slug');
+const typeTitle = require('./type-title');
+const pathDiff = require('./path-diff');
+const {
+    addTrailingSlash,
+    removeTrailingSlash,
+    addLeadingSlash,
+    removeLeadingSlash,
+} = require('./path-slashes');
```
Comment (gingermusketeer) [inline lib/helpers/index.js:13]: With ESM modules not supporting the index.js pattern would it make sense to avoid using it so that migrating to ESM modules is easier in the future?

File: lib/helpers/path-diff.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// @ts-check
+/* eslint-disable no-continue */
+const { basename, normalize } = require('path');
+const isGlob = require('is-glob');
+const { addLeadingSlash } = require('./path-slashes');
+
+/**
+ * Takes a definition string and uses it to remove leading parts of a path if necessary.
+ * The definition may include glob segments which, if present, will be removed before use from the definition along with all trailing parts of the definition string.
```
Comment (gingermusketeer) [inline lib/helpers/path-diff.js:9]: Might be nice to wrap this line as it is a bit long for GitHub.

File: lib/helpers/path-slashes.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+const addLeadingSlash = (val) => (val.startsWith('/') ? val : `/${val}`);
+
+/**
+ * Remove a leading slash from a path if necessary
+ * 
+ * @param {string} val
+ *
+ * @returns {string}
+ */
+const removeLeadingSlash = (val) => (val.startsWith('/') ? val.substr(1) : val);
```
Comment (gingermusketeer) [inline lib/helpers/path-slashes.js:36]: This function isn't being used anywhere?

File: lib/helpers/type-title.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// @ts-check
+
+module.exports = (type) => {
+    if (type === 'package') return 'PACKAGE';
+    if (type === 'npm') return 'NPM';
+    return 'MAP';
```
Comment (gingermusketeer) [inline lib/helpers/type-title.js:6]: Would `type.toUpperCase()` work here?

File: test/classes/eik-config.test.js [hunk 9]
```diff
@@ -291,3 +294,507 @@ test('pathsAndFilesAbsolute - does not throw for directory glob pattern', async
...
+    t.equal(mappings[1].source.relative, 'folder/styles.css');
+    t.match(mappings[1].source.absolute, 'test/fixtures/folder/styles.css');
+    t.equal(mappings[1].destination.packagePathname, '/pkg/pizza/0.0.0');
+    t.equal(mappings[1].destination.filePathname, '/styles.css');
+    t.equal(
+        mappings[1].destination.url.href,
+        'http://server/pkg/pizza/0.0.0/styles.css',
+    );
+    t.end();
+});
```
Comment (gingermusketeer) [inline test/classes/eik-config.test.js:450]: Is this test a duplicate of the test two up starting at line 356?

File: test/helpers/path-diff.test.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+tap.test('definition: ./**/*, path: /folder/client.js => /folder/client.js', (t) => {
+    t.equal(pathDiff('./**/*', '/folder/client.js'), '/folder/client.js');
+    t.end();
+});
+
+tap.test('definition: ./**/*, path: /client.js => /client.js', (t) => {
+    t.equal(pathDiff('./**/*', '/client.js'), '/client.js');
+    t.end();
+});
```
Comment (gingermusketeer) [inline test/helpers/path-diff.test.js:37]: Is this needed? Seems like the same as the one above.

File: test/helpers/path-diff.test.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+tap.test('definition: ./*, path: /client.js => /client.js', (t) => {
+    t.equal(pathDiff('./*', '/client.js'), '/client.js');
+    t.end();
+});
+
+tap.test('definition: path/to/folder, path: /path/to/folder/to/folder/client.js => /to/folder/client.js', (t) => {
+    t.equal(pathDiff('path/to/folder', 'path/to/folder/to/folder/client.js'), '/to/folder/client.js');
+    t.end();
+});
```
Comment (gingermusketeer) [inline test/helpers/path-diff.test.js:47]: This seems the same as the first test?

File: test/helpers/path-diff.test.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+tap.test('definition: /folder, path: /path/to/folder/to/folder/client.js => /path/to/folder/to/folder/client.js', (t) => {
```
Comment (gingermusketeer) [inline test/helpers/path-diff.test.js:49]: Is this a likely case?

File: test/helpers/path-diff.test.js [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    t.equal(pathDiff('', './client.js'), '/client.js');
+    t.end();
+});
+
+tap.test('definition: null, path: ./client.js => /client.js', (t) => {
+    t.equal(pathDiff(null, './client.js'), '/client.js');
+    t.end();
+});
+
+tap.test('definition: undefined, path: ./client.js => /client.js', (t) => {
```
Comment (gingermusketeer) [inline test/helpers/path-diff.test.js:84]: The test descriptions are basically identical to the assertion in the test. Perhaps it is worth being a bit more high level? `handles prefix being undefined` for example.

File: test/fixtures/folder/client.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+'use strict';
```
Comment (gingermusketeer) [inline test/fixtures/folder/client.js:1]: Do the fixtures need to be javascript files? Might be nice to avoid having JS so that the NPM script doesn't need to be changed and a single test can be run with `npm run test -- [some test file]`

File: test/helpers/config-store.test.js [hunk 15]
```diff
@@ -221,24 +221,3 @@ test('saves config to disk - invalid config', async (t) => {
...
-    try {
-        configStore.persistToDisk(config);
-    } catch (e) {
-        t.equal(
-            e.message,
-            'The "path" argument must be of type string. Received undefined',
-        );
-    }
-    t.end();
-});
```
Comment (gingermusketeer) [inline test/helpers/config-store.test.js:244]: Isn't this still possible now that typescript checking has been added? If a consuming library isn't using typescript then it won't have this check provided.

File: test/classes/eik-config.test.js [hunk 16]
```diff
@@ -291,3 +294,507 @@ test('pathsAndFilesAbsolute - does not throw for directory glob pattern', async
...
+    t.match(mappings[1].source.absolute, 'test/fixtures/folder/styles.css');
+    t.equal(mappings[1].destination.packagePathname, '/pkg/pizza/0.0.0');
+    t.equal(mappings[1].destination.filePathname, '/path/to/folder/styles.css');
+    t.equal(
+        mappings[1].destination.url.href,
+        'http://server/pkg/pizza/0.0.0/path/to/folder/styles.css',
+    );
+
+    t.end();
+});
```
Comment (gingermusketeer) [inline test/classes/eik-config.test.js:800]: Some of these tests are perhaps checking the behaviour of `resolvedFiles` method? Perhaps it is worth testing that function directly?

Also I think it would be good to test the error cases for the mapping function as I assume the other methods which have similar logic will be removed along with the tests in the future.

**Review by gingermusketeer — COMMENTED** [review]
I see a few warnings related to missing newlines. Perhaps good to get prettier setup here?

File: eikjson.d.ts [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    | string;
+  /**
+   * Import map files given by URL(s) to be used during package bundling. Specified as a URL or array of URLs. URLs are locations of import map files that follow the W3C import map spec.
+   */
+  "import-map"?: string | string[];
+  /**
+   * Location of local Eik build directory (defaults to .eik). Used by Eik client.
+   */
+  out?: string;
+}
```
[the remark this answers is not in the document]
Comment (digitalsadhu) [reply eikjson.d.ts:41]: that's a good suggestion 👍

File: lib/classes/eik-config.js (hunk 3 above)
Comment (digitalsadhu) [reply lib/classes/eik-config.js:130]: yeah, I think that they probably would

File: lib/classes/eik-config.js (hunk 4 above)
Comment (digitalsadhu) [reply lib/classes/eik-config.js:174]: Yea, probably should do something here true. I could also just delete them now that I have things working and this release is super breaking anyway maybe

File: lib/helpers/index.js (hunk 5 above)
Comment (digitalsadhu) [reply lib/helpers/index.js:13]: I don't see it as a huge deal I don't think, you can still do

```js
import pathSlashes from './path-slashes';
const { addLeadingSlash } = pathSlashes;
```

Most likely I would migrate the path slashes module itself to esm and it would work on one line. The alternative is that I would need to have a file for every function which would result in multiple import statements right? Anything I'm missing in my thinking here?

File: test/helpers/path-diff.test.js (hunk 12 above)
Comment (digitalsadhu) [reply test/helpers/path-diff.test.js:49]: Nope, just wanted to avoid any unexpected weirdness so tried to think of some very corner cases and how they should be handled

File: lib/helpers/index.js (hunk 5 above)
Comment (gingermusketeer) [reply lib/helpers/index.js:13]: Sorry I meant when loading importing not when exporting from this file. With ESM you would have to do `import {addLeadingSlash} from "../helpers/index.js"` might as well go `import {addLeadingSlash} from "../helpers/path-slashes.js"`?

File: lib/classes/eik-config.js (hunk 4 above)
Comment (gingermusketeer) [reply lib/classes/eik-config.js:174]: I guess it is only breaking if you use the new function? Could keep the old ones around until everything has been migrated. Though not sure if that helps. Might be nicer to see breaking tests rather than have everything working.

File: lib/helpers/index.js (hunk 5 above)
Comment (digitalsadhu) [reply lib/helpers/index.js:13]: ah true. I should probably use package exports to set up the various helpers and such the common package exposes

File: lib/classes/eik-config.js (hunk 4 above)
Comment (digitalsadhu) [reply lib/classes/eik-config.js:174]: yeah, kinda tempted to go hard breaking here and rip off the bandage.

File: lib/helpers/index.js (hunk 5 above)
Comment (trygve-lie) [reply lib/helpers/index.js:13]: I would not worry too much about this at the moment. I would like a separate task on reworking the public API of this module. The current public API is a bit confusing (probably due to a lot of the code being moved here from the internals of the client). 

Ex: to load the eik config in the plugins we're now calling `helpers.getDefaults(path)`. That's not really a well described method for loading a config.

File: test/fixtures/folder/client.js (hunk 14 above)
Comment (trygve-lie) [reply test/fixtures/folder/client.js:1]: For single files one normally just run `node path/to/file.js` with tap. Tap differ here from tools like jest and mocha.

Comment (gingermusketeer) [reply test/fixtures/folder/client.js:1]: I tried that but the output is a bit hard to read compared to using the tap command.

Comment (digitalsadhu) [reply test/fixtures/folder/client.js:1]: I've fixed this in a different way. `tap --test-regex=.test.js --test-ignore=node_modules`
Now you can do `npm test` or `npm test -- test/classes/eik-config.test.js` as you like.

File: test/helpers/config-store.test.js (hunk 15 above)
Comment (digitalsadhu) [reply test/helpers/config-store.test.js:244]: I think the test was verifying that there was a cwd path which now defaults to process.cwd so the persistToDisk method no longer threw for this case

**Comment by trygve-lie** [comment]
This is breaking; should we do a `next` branch and merge this in to it and cut a `next` release from there?

@gingermusketeer What do you think?

**Comment by digitalsadhu** [comment]
sounds good to me

**Comment by trygve-lie** [comment]
I made a `next` branch.

File: test/helpers/config-store.test.js (hunk 15 above)
Comment (gingermusketeer) [reply test/helpers/config-store.test.js:244]: Ah okay I missed that change. Good to remove this in that case.

**Comment by trygve-lie** [comment]
The `next` branch should now cut a `next` pre released when merged into it. Lets see if that happens.
